File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        Map<Character, Character> map = Map.of(')', '(', '}', '{', ']', '[');
        Deque<Character> stk = new ArrayDeque<>();

        for (char bracket: s.toCharArray()) {
            if (map.containsKey(bracket)) {
                if (map.get(bracket) == stk.peek()) {
                    stk.pop();
                } else {
                    return false;
                }
            } else {
                stk.push(bracket);
            }
        }

        if (stk.size() != 0) {
            return false;
        }
        return true;
    }
}

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    inline bool isValid(string s) {
        int n = s.size();
        unordered_map<char, char> map = {{')', '('}, {'}', '{'}, {']', '['}};
        
        stack<char> st;
        for (int i = 0; i < n; i++) {
            if (map.count(s[i])) {
                if (!st.empty() && st.top() == map[s[i]]) {
                    st.pop();
                } else {
                    return false;
                }
            } else {
                st.push(s[i]);
            }
        }

        return st.empty();
    }
};